design/20221205-memory-management.md

```diff
@@ -0,0 +1,454 @@
+
+This should reduce the excessive memory consumption caused by caching full contents of cert-manager unrelated `Secret`s whilst still ensuring that most of the `Secret`s that cert-manager needs frequently are retrieved from cache.
+
+### Background
+
+The excessive memory consumption comes from the amount of cluster objects being stored in the [shared informers cache](https://github.com/kubernetes/client-go/blob/v12.0.0/tools/cache/shared_informer.go#L47-L58).
+cert-manager uses client-go's [informer factory](https://github.com/kubernetes/client-go/tree/master/informers) to create informers for core types. We have [auto-generated informers](https://github.com/cert-manager/cert-manager/tree/v1.10.1/pkg/client/informers/externalversions) for cert-manager.io types. These informers do not directly expose the cache or the [ListerWatcher](https://github.com/kubernetes/client-go/blob/v12.0.0/tools/cache/shared_informer.go#L188) which is responsible for listing and setting up watches for objects.
+
+### Design details
+
+Have two informers that watch `Secret` objects.
+
+One would be a [metadata only](#partial-object-metadata) informer and would watch all `Secret` objects that _don't_ have a cert-manager label. This informer would list, watch and cache metadata of all `Secret`s that don't have a cert-manager label.
+
+The other would be a typed informer like we already have, but with a filter to watch only `Secret` objects that _do_ have a cert-manager label.
+
+The reconcile loops would have two listers (for metadata and typed informer cache) and a live client to get `Secret`s from kube apiserver directly.
+
+In practice this would mean that:
+- cert-manager would cache the full `Secret` object for all `certificate.spec.secretName` `Secret`s and all `Secret`s containing temporary private keys in almost all cases and would retrieve these `Secret`s from cache in almost all cases (see the section about [Secrets for Certificates](#Secrets-for-Certificates))
+
+- cert-manager would cache only `PartialObjectMetadata` for all `Secret`s that are used by issuers/cluster-issuers (account keys, tokens etc) and would retrieve the full contents of those `Secret`s from kube-apiserver - unless users would label them with a `cert-maanger.io` label in which case they would be stored in cache in full and retrieved from there
+
+- cert-manager would cache `PartialObjectMetadata` for all other unrelated `Secret`s
+
+[Here](https://github.com/irbekrm/cert-manager/tree/partial_metadata) is an experimental 'partial_metadata' branch that implements a prototype:
+[Two informer factories](https://github.com/irbekrm/cert-manager/blob/partial_metadata/pkg/controller/context.go#L293-L297) have been set up for this controller, one a partial metadata one with a selector to only watch unlabelled resources, one to watch Secret types with a label filter.
```

Comment:
The design details don't give enough information about how the two informers will be used.
Paste the code from your prototype here so that we can comment on it directly.

Reply:
The code of `SecretGetter` I mean.

Reply:
Thanks Richard, that's a good idea.

I added implementation section here, let me know if that makes it more clear https://github.com/irbekrm/cert-manager/blob/memory_design/design/20221205-memory-management.md#implementation